File: greedy_algorithms/fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <stdint.h>
#include <inttypes.h>
#include <math.h> 
#include <iomanip>

using namespace std;

int fractional_knapsack(int capacity, vector<int>&weights, vector<int>&values){
	// retrieve the number of different objects that can be included in the knapsack
	int n = weights.size();
	// track the value of the objects that have been included in the knapsack
	double added_value = 0.0;
	// loop over the objects
	for(int i = 0; i < n; ++i){
		// terminate the algorithm if there is no capacity left (i.e., knapsack is full)
		if (capacity == 0){
			// return the value of the objects that have been included in the knapsack
			return added_value;
		}
		// determine the current optimal choice
		double optimal_choice = 0.0;
		int optimal_choice_idx = 0;
		for(int j = 0; j < n; ++j){
			if((values[j]/ (double) weights[j] > optimal_choice) & (weights[j] > 0)){
				optimal_choice = values[j]/ (double) weights[j];
				optimal_choice_idx = j;
			}
	    }
        // add the optimal choice to the knapsack and update the weights and current capacity
	    double current_pick = min(weights[optimal_choice_idx], capacity);
	    added_value += current_pick*optimal_choice;
	    weights[optimal_choice_idx] -= current_pick;
	    capacity -= current_pick;		
	}
	// return the value of the objects that have been included in the knapsack
	return added_value;
}

int main() {
  int n;
  int capacity;
  cout << "Input the number of different objects that can be added to the knapsack (integer): ";
  cin >> n ;
  cout << "Input the capacity of the knapsack (integer): ";
  cin >> capacity;
  vector<int> values(n);
  vector<int> weights(n);
  cout << "Input the value and weight of each object that can be added to the knapsack (integer + space + integer): " << "\n";
  for (int i = 0; i < n; i++) {
      cin >> values[i] >> weights[i];
  }

  double value_in_knapsack = fractional_knapsack(capacity, weights, values);

  cout << "The maximum attainable value that can be added to the knapsack is: ";
  cout << std::setprecision(10) << value_in_knapsack << endl;
  return 0;
}
	

File: greedy_algorithms/fractional_knapsack.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jan  6 01:26:53 2021

@author: Alberto Ciacci
"""
import numpy as np

def fractional_knapsack(capacity, values, weights):
    
    picked_values  = np.zeros(n)
    picked_weights = np.zeros(n)
    for i in range(n):
        if capacity == 0:
           return picked_values, picked_weights
        best_score = 0
        best_idx = 0
        for j in range(n):
            if weights[j] > 0:
                current_score = values[j]/float(weights[j])
                if (current_score > best_score):
                    best_score = current_score
                    best_idx = j
        picked_weight = np.minimum(weights[best_idx],capacity)
        weights[best_idx] -= picked_weight
        capacity -= picked_weight
        picked_weights[best_idx] += picked_weight
        picked_values[best_idx] += picked_weight*best_score
    return picked_values, picked_weights
            

if __name__ == "__main__":
    
    n = int(input("Input the number of different objects that can be added to the knapsack (integer): "))
    capacity = int(input("Input the capacity of the knapsack (integer): "))
    values   = np.zeros(n)
    weights  = np.zeros(n)
    for i in range(n):
        values[i] = int(input("Input the value of the {0}-th object: ".format(i+1)))
        weights[i] = int(input("Input the weight of the {0}-th object: ".format(i+1)))
    picked_values, picked_weights = fractional_knapsack(capacity, values, weights)
    print("The fractional knapsack algorithm selected the following objects: ")
    for i in range(n):
        print("object: {0}, weight: {1}, value: {2}".format(i+1,picked_weights[i],picked_values[i]))
    print("The total value of the knapsack is: {0}".format(np.sum(picked_values)))